cpp/oneapi/dal/graph/undirected_adjacency_array_graph.hpp

```diff
@@ -114,30 +64,89 @@ class ONEDAL_EXPORT undirected_adjacency_array_graph {
 
     /// Constructs an empty undirected_adjacency_array_graph with specified graph properties
     /// and allocator
-    undirected_adjacency_array_graph(const graph_user_value_type &graph_user_value,
-                                     allocator_type allocator = allocator_type()){};
+    undirected_adjacency_array_graph(const GraphValue &graph_user_value,
+                                     Allocator allocator = Allocator()){};
 
     /// Constructs an empty undirected_adjacency_array_graph with move graph properties and
     /// allocator
-    undirected_adjacency_array_graph(graph_user_value_type &&graph_user_value,
-                                     allocator_type allocator = allocator_type()){};
+    undirected_adjacency_array_graph(graph_user_value_type<graph_type> &&graph_user_value,
+                                     Allocator allocator = Allocator()){};
 
     /// Copy operator for undirected_adjacency_array_graph
     undirected_adjacency_array_graph &operator=(const undirected_adjacency_array_graph &graph);
 
     /// Move operator for undirected_adjacency_array_graph
     undirected_adjacency_array_graph &operator=(undirected_adjacency_array_graph &&graph);
-    using pimpl =
-        oneapi::dal::detail::pimpl<detail::undirected_adjacency_array_graph_impl<VertexValue,
-                                                                                 EdgeValue,
-                                                                                 GraphValue,
-                                                                                 IndexType,
-                                                                                 Allocator>>;
 
 private:
+    using pimpl = oneapi::dal::detail::pimpl<typename graph_traits<graph_type>::impl_type>;
+
     pimpl impl_;
+
     friend pimpl &oneapi::dal::preview::detail::get_impl<graph_type>(graph_type &graph);
 
     friend const pimpl &oneapi::dal::preview::detail::get_impl<graph_type>(const graph_type &graph);
 };
+
+template <typename VertexValue,
+          typename EdgeValue,
+          typename GraphValue,
+          typename IndexType,
+          typename Allocator>
+struct graph_traits<
+    undirected_adjacency_array_graph<VertexValue, EdgeValue, GraphValue, IndexType, Allocator>> {
+    using graph_type =
+        undirected_adjacency_array_graph<VertexValue, EdgeValue, GraphValue, IndexType, Allocator>;
+    using impl_type = detail::undirected_adjacency_array_graph_impl<VertexValue,
```

Comment:
Is it part of public interface available to the user? If so, do you really want to expose API for getting underlying graph impl type?

Reply:
> Is it part of public interface available to the user? If so, do you really want to expose API for getting underlying graph impl type?
We expose defined by impl types here not to expose them in graph. You can see that a lot of things are removed from public class. Concretely this using is service with the purpose not to duplicate long name undirected_adjacency_array_graph<VertexValue, EdgeValue, GraphValue, IndexType, Allocator>.

Reply:
Do you expect users to use `graph_traits` directly in their code?

Reply:
Expected. The goal of graph traits is to provide the opportunity for adaption of the external graph types. That is the user will create another one specialization of graph traits concretely for its graph.

Reply:
Well, do you want user to know type of the `impl` class hidden behind `undirected_adjacency_array_graph` or any other type?

Reply:
It is not only for the get_impl. using vertex_set = typename graph_traits<Graph>::vertex_set; in graph loading. Moreover vertex_size_type is based on its set type and it is anyway exposed from the impl. To avoid duplication, it is exposed directly in graph traits.